test/methods.jl

```diff
@@ -1,30 +1,37 @@
 using FiniteDifferences: Forward, Backward, Central, Nonstandard
 
 @testset "Methods" begin
-    for f in [:forward_fdm, :backward_fdm, :central_fdm]
-        @eval @test $f(1, 0; bound=1)(sin, 1) == sin(1)
-        @eval @test $f(2, 0; bound=1)(sin, 1) == sin(1)
-        @eval @test $f(3, 0; bound=1)(sin, 1) == sin(1)
-        @eval @test $f(10, 1; bound=1)(sin, 1) ≈ cos(1)
-        @eval @test $f(10, 2; bound=1)(sin, 1) ≈ -sin(1)
-
-        @eval @test $f(1, 0; bound=1)(exp, 1) == exp(1)
-        @eval @test $f(2, 0; bound=1)(exp, 1) == exp(1)
-        @eval @test $f(3, 0; bound=1)(exp, 1) == exp(1)
-        @eval @test $f(10, 1; bound=1)(exp, 1) ≈ exp(1)
-        @eval @test $f(10, 2; bound=1)(exp, 1) ≈ exp(1)
-
-        @eval @test $f(1, 0; bound=1)(abs2, 1) == 1
-        @eval @test $f(2, 0; bound=1)(abs2, 1) == 1
-        @eval @test $f(3, 0; bound=1)(abs2, 1) == 1
-        @eval @test $f(10, 1; bound=1)(abs2, 1) ≈ 2
-        @eval @test $f(10, 2; bound=1)(abs2, 1) ≈ 2
-
-        @eval @test $f(1, 0; bound=1)(sqrt, 1) == 1
-        @eval @test $f(2, 0; bound=1)(sqrt, 1) == 1
-        @eval @test $f(3, 0; bound=1)(sqrt, 1) == 1
-        @eval @test $f(10, 1; bound=1)(sqrt, 1) ≈ .5
-        @eval @test $f(10, 2; bound=1)(sqrt, 1) ≈ -.25
+
+    # The different approaches to approximating the gradient to try.
+    methods = [forward_fdm, backward_fdm, central_fdm]
+    methods = [forward_fdm]
+
+    # The different floating-point types to try.
+    types = [Float32, Float64]
+
+    # The different functions to evaluate (.f), their first derivative at 1 (.d1),
+    # and second derivative at 1 (.d2).
+    foos = [
+        (f=sin, d1=cos(1), d2=-sin(1)),
+        (f=exp, d1=exp(1), d2=exp(1)),
+        (f=abs2, d1=2, d2=2),
+        (f=sqrt, d1=0.5, d2=-0.25),
+    ]
+
+    # Test all combinations of the above settings. i.e. differentiate all functions using
+    # all methods and data types.
+    @testset "foo=$(foo.f), method=$m, type=$T" for foo in foos, m in methods, T in types
+
+        @testset "method-order=$order" for order in [1, 2, 3]
+            @test m(order, 0; bound=1)(foo.f, T(1)) isa T
+            @test m(order, 0; bound=1)(foo.f, T(1)) == T(foo.f(1))
+        end
+
+        @test m(10, 1; bound=1)(foo.f, T(1)) isa T
+        @test m(10, 1; bound=1)(foo.f, T(1)) ≈ T(foo.d1)
+
+        @test m(10, 2; bound=1)(foo.f, T(1)) isa T
+        @test m(10, 2; bound=1)(foo.f, T(1)) ≈ T(foo.d2)
```

Comment:
Hmm, this line fails for `Float32`s in combination with `forward_fdm` and `backward_fdm`, but it succeeds for `central_fdm`.

I removed the converts inside the `fdm` function, and that fixed it. Would we be happy with performing the step size and derivative estimation at the highest possible accuracy and only converting back to `T` when returning the answser?

There is one more issue with the tests, which is that `backward_fdm` will produce numbers outside the domain of `sqrt`. This can be fixed by using `(f=x -> sqrt(x + 1), d1=0.5 / sqrt(2), d2=-0.25 / 2^(3/2))` instead.

I've branched off and pushed the above proposed fixes to `wb/add-tiny-fixes`.

Reply:
Hmm operating at the highest accuracy available definitely sounds appealing.

> I removed the converts inside the fdm function, and that fixed it. Would we be happy with performing the step size and derivative estimation at the highest possible accuracy and only converting back to T when returning the answser?

I think that we want to be careful here. In particular, we probably want to make sure that we're actually taking the correct code path, in the sense that we should make sure we're actually evaluating `f` using `Float32`s if we're trying out `Float32`s.

I'm imagining a situation in which the `Float32` implementation of `f` has slightly different characteristics from the `Float64` implementation and, as such, finite-differencing the `Float64` implementation gives the "wrong" answer. I'm not sure that I have a good answer off the top of my head though.

I think I would generally like to think a bit about the consequences of evaluating the function at a different level of precision before we commit to it, specifically the difference between different floating-point precision levels as opposed to `Integer` vs float, which we know causes issues for e.g. exponentiation -- it might be that it's totally fine.

Reply:
> I'm imagining a situation in which the Float32 implementation of f has slightly different characteristics from the Float64 implementation and, as such, finite-differencing the Float64 implementation gives the "wrong" answer. I'm not sure that I have a good answer off the top of my head though.

Ah, yes, that's definitely something to get right. I might have overlooked that in my branch. Let me get that right and what happens.

At a first thought, I would be in favour of providing the same data type to the function as the point at which we are evaluating the derivative.

Reply:
Yeah, that's my thinking too. I can see nothing wrong with doing step-size calculations for operations involving low-precision at higher-precision.

Reply:
The tests then fail. Interesting. Do you have any thoughts on this? Perhaps using forward differences at `Float32` to estimate a second derivative is pushing the method too hard. I'm not sure.

Reply:
Yeah, maybe we just shouldn't be testing that 🤷‍♂️ . @oxinabox any thoughts?

Reply:
I do not have any thoughts.
I will point out that the `widen` function exists. 
```
julia> widen(Float32)
Float64
```


Maybe @shashi  has opinions, @shashi has done work on finite differencing tools before

Reply:
In the absence of other immediate options, I've stopped testing the second-derivative stuff on `Float32`s.